lib/rouge/lexer.rb

```diff
@@ -501,12 +506,45 @@ def self.detect?(text)
   end
 
   module Lexers
+    ProxyLexer = Struct.new(:path, :name, :tag, :aliases, :filenames, :mimetypes) do
+      def method_missing(name, *args, &blk)
+        Lexers.load_lexer self.path
+        lexer = Lexers.const_get self.name
+        lexer.send(name, *args, &blk)
+      end
+    end
+
     @_loaded_lexers = {}
+    @_loaded_proxies = {}
 
     def self.load_lexer(relpath)
       return if @_loaded_lexers.key?(relpath)
       @_loaded_lexers[relpath] = true
       load File.join(__dir__, 'lexers', relpath)
     end
+
+    def self.load_proxies(cache_file)
+      registry = {}
+
+      JSON.parse(File.read(cache_file)).each do |h|
+        proxy = ProxyLexer.new(h["path"],
+                               h["name"],
+                               h["tag"],
+                               h["aliases"] || [],
+                               h["filenames"] || [],
+                               h["mimetypes"] || [])
+        @_loaded_proxies[proxy.name] = proxy
+        registry[proxy.tag] = proxy
+        proxy.aliases.each { |a| registry[a] = proxy } unless proxy.aliases.nil?
+      end
+
+      Rouge::Lexer.load_registry registry
+    end
+
+    def self.const_missing(name)
+      raise NameError unless (proxy = @_loaded_proxies[name.to_s])
```

Comment:
Not sure if `name.to_s` here will lead to extraneous String allocations.

Reply:
I'm not sure it matters whether it does or not. The type of the value given to `const_missing` is decided by Ruby. There's no way around transforming it to a String.